src/Lecture2.hs

```diff
@@ -155,12 +178,70 @@ You're free to define any helper functions.
 
 -- some help in the beginning ;)
 data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
-    }
-
-dragonFight = error "TODO"
+  { knightHealth    :: Int
+  , knightAttack    :: Int
+  , knightEndurance :: Int
+  }
+
+knightTurn :: (Knight, Dragon) -> (Knight, Dragon)
+knightTurn (knight, dragon) =
+  (knight {knightEndurance = (knightEndurance knight) - 1}
+  ,dragon {dragonHealth = (dragonHealth dragon) - (knightAttack knight)})
+
+data Treasure = Treasure
+  deriving (Show, Eq)
+data Chest = Chest
+  { chestGold     :: Natural
+  , chestTreasure :: Maybe Treasure
+  }
+  deriving (Show, Eq)
+
+data DragonType = Red | Black | Green
+  deriving (Show, Eq)
+
+getRandomGoldAmount :: Natural
+getRandomGoldAmount = 4 -- todo, randomize
+
+getDragonReward :: DragonType -> DragonFightResult
+getDragonReward Green = Reward (Chest getRandomGoldAmount Nothing) 250
+getDragonReward Red = Reward (Chest getRandomGoldAmount (Just Treasure)) 100
+getDragonReward Black = Reward (Chest getRandomGoldAmount (Just Treasure)) 150
+
+data Dragon = Dragon
+  { dragonHealth    :: Int
+  , dragonType      :: DragonType
+  , dragonAttack    :: Int
+  }
+  deriving (Show, Eq)
+
+dragonTurn :: Int -> (Knight, Dragon) -> (Knight, Dragon)
+dragonTurn turnNumber (knight, dragon)
+  | turnNumber == 10 =
+      (knight {knightHealth = (knightHealth knight) - (dragonAttack dragon)}
+      ,dragon)
+  | otherwise = (knight, dragon)
+
+-- what happens if knight and dragon kill each other simultaneously?
+-- I want to assume the guards are parsed in order, so Reward is recognized only if first
+-- two aren't true.
+doRound :: Int -> (Knight, Dragon) -> DragonFightResult
+doRound roundNum (knight, dragon)
+  | knightHealth knight    <= 0 = Death
+  | knightEndurance knight <= 0 = Retreat
+  | dragonHealth dragon    <= 0 = (getDragonReward (dragonType dragon))
+  | otherwise = 
+      doRound (roundNum + 1) (dragonTurn (roundNum + 1) (knightTurn (knight, dragon)))
+
+data DragonFightResult = Death
+                       | Retreat
+                       | Reward
+                         { chest :: Chest
+                         , exp  :: Natural -- assuming like gold
+                         }
+  deriving (Show, Eq) 
+  
+dragonFight :: Knight -> Dragon -> DragonFightResult
+dragonFight knight dragon = doRound 0 (knight, dragon)
```

Comment:
Is there a reason to put `knight` and `dragon` inside the pair? 🤔 
If you don't do this, you can eta-reduce your function:

```suggestion
dragonFight = doRound 0
```